File: benchmark/quicksort/sort.py
def swap(arr, i, j):
    arr[i], arr[j] = arr[j], arr[i]


def partition(arr, low, high):
    pivot = arr[high]
    i = low - 1
    for j in range(low, high):
        if arr[j] <= pivot:
            i += 1
            swap(arr, i, j)
    swap(arr, i + 1, high)
    return i + 1


def quicksort(arr, low, high):
    if low < high:
        pi = partition(arr, low, high)
        quicksort(arr, low, pi - 1)
        quicksort(arr, pi + 1, high)


def is_sorted(arr):
    for i in range(1, len(arr)):
        if arr[i - 1] > arr[i]:
            return False
    return True


def main():
    size = 3000
    iterations = 20
    checksum = 0

    for iter in range(iterations):
        # Create array with pseudo-random values
        arr = [0] * size
        for i in range(size):
            # Simple LCG: (a * x + c) mod m
            arr[i] = ((i * 1103515245 + 12345 + iter) % 1000000) - 500000

        quicksort(arr, 0, size - 1)

        if is_sorted(arr):
            checksum += arr[0] + arr[size - 1]

    print(checksum)


import sys
sys.setrecursionlimit(20000)
main()

File: benchmarks/quicksort/sort.js
function swap(arr, i, j) {
    const tmp = arr[i];
    arr[i] = arr[j];
    arr[j] = tmp;
}

function partition(arr, low, high) {
    const pivot = arr[high];
    let i = low - 1;
    for (let j = low; j < high; j++) {
        if (arr[j] <= pivot) {
            i++;
            swap(arr, i, j);
        }
    }
    swap(arr, i + 1, high);
    return i + 1;
}

function quicksort(arr, low, high) {
    if (low < high) {
        const pi = partition(arr, low, high);
        quicksort(arr, low, pi - 1);
        quicksort(arr, pi + 1, high);
    }
}

function isSorted(arr) {
    for (let i = 1; i < arr.length; i++) {
        if (arr[i - 1] > arr[i]) {
            return false;
        }
    }
    return true;
}

const size = 3000;
const iterations = 200;
let checksum = 0;

for (let iter = 0; iter < iterations; iter++) {
    const arr = new Array(size);
    for (let i = 0; i < size; i++) {
        arr[i] = ((i * 1103515245 + 12345 + iter) % 1000000) - 500000;
    }

    quicksort(arr, 0, size - 1);

    if (isSorted(arr)) {
        checksum += arr[0] + arr[size - 1];
    }
}

console.log(checksum);
